sigstore/_internal/trustroot.py

```diff
@@ -58,48 +76,201 @@ def _is_timerange_valid(period: TimeRange | None, *, allow_expired: bool) -> boo
+        the signature is invalid.
+        """
+        key = self._keyring.get(key_id)
+        if key is None:
+            # If we don't have a key corresponding to this key ID, we can't
+            # possibly verify the signature.
+            raise KeyringLookupError(f"no known key for key ID {key_id.hex()}")
+
+        try:
+            if isinstance(key, rsa.RSAPublicKey):
+                key.verify(
+                    signature=signature,
+                    data=data,
+                    padding=padding.PKCS1v15(),
+                    algorithm=hashes.SHA256(),
+                )
+            elif isinstance(key, ec.EllipticCurvePublicKey):
+                key.verify(
+                    signature=signature,
+                    data=data,
+                    signature_algorithm=ec.ECDSA(hashes.SHA256()),
+                )
+            else:
+                # NOTE(ww): Unreachable without API misuse.
+                raise KeyringError(f"unsupported key type: {key}")
+        except InvalidSignature as exc:
+            raise KeyringSignatureError("invalid signature") from exc
+
+
+RekorKeyring = NewType("RekorKeyring", Keyring)
+CTKeyring = NewType("CTKeyring", Keyring)
+
+
+class KeyringPurpose(str, Enum):
+    """
+    Keyring purpose typing
+    """
+
+    SIGN = "sign"
+    VERIFY = "verify"
+
+    def __str__(self) -> str:
+        """Returns the purpose string value."""
+        return self.value
+
+
 class TrustedRoot(_TrustedRoot):
     """Complete set of trusted entities for a Sigstore client"""
 
+    args: Optional[Namespace] = None
+    purpose: KeyringPurpose
+
     @classmethod
-    def from_file(cls, path: str) -> "TrustedRoot":
+    def from_file(
+        cls,
+        path: str,
+        args: Optional[Namespace] = None,
+        purpose: KeyringPurpose = KeyringPurpose.VERIFY,
+    ) -> "TrustedRoot":
         """Create a new trust root from file"""
-        tr: TrustedRoot = cls().from_json(Path(path).read_bytes())
-        return tr
+        trusted_root: TrustedRoot = cls().from_json(Path(path).read_bytes())
+        trusted_root.args = args
+        trusted_root.purpose = purpose
+        return trusted_root
 
     @classmethod
-    def from_tuf(cls, url: str, offline: bool = False) -> "TrustedRoot":
+    def from_tuf(
+        cls,
+        url: str,
+        offline: bool = False,
+        args: Optional[Namespace] = None,
+        purpose: KeyringPurpose = KeyringPurpose.VERIFY,
+    ) -> "TrustedRoot":
         """Create a new trust root from a TUF repository.
 
         If `offline`, will use trust root in local TUF cache. Otherwise will
         update the trust root from remote TUF repository.
         """
         path = TrustUpdater(url, offline).get_trusted_root_path()
-        return cls.from_file(path)
+        trusted_root = cls.from_file(path)
+        trusted_root.args = args
+        trusted_root.purpose = purpose
+        return trusted_root
```

Comment:
```suggestion
        return cls.from_file(path, args, purpose)
```

same suggestion applies to all other heper constructors